### PR TITLE
Feat(build-manager): introduce a persistent state for build jobs

### PR DESCRIPTION
**Pull request by abrl91**
Close: #7393

### Summary
📦🌐🚀

This pull request adds Redis support to the `build-manager` service and improves the code generation process. It adds a `redis` service and a `redis_data` volume to the `docker-compose.dev.yml` file. It also adds the `ioredis` package as a dependency and defines a `RedisModule` and a `RedisService` to handle the Redis connection and operations. It updates the `build-runner` and the `code-generator-splitter` services to use imported types and to set and get the job status in Redis using the `buildIdWithDomainName` as the key. It also updates the test file and the environment variables for the `build-manager` service.

> _We summon the power of Redis_
> _To cache and message our progress_
> _We generate code with domain names_
> _We split and build with no shame_

### Walkthrough
*  Add Redis service and connection to `docker-compose.dev.yml` and `redis.module.ts` ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-9542f82d64bbeebd91f6236324bfe199e9657e2cb1fd9779d5d6dcdcf9cd4de1R19-R25), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-9542f82d64bbeebd91f6236324bfe199e9657e2cb1fd9779d5d6dcdcf9cd4de1R74), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-33c9cee79af2554b663b569cf8f1f3014ab3c4c68846c9b320abf8048e56a0cdR1-R28))
*  Add `ioredis` dependency to `package.json` ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R122))
*  Add environment variables for Redis connection to `.env` and `env.ts` ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-ead4ab1598bcdb2b9593a43f5299b26787f4709777e29d06535eeca52d411333R9-R14), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-7414ed5c548aedbceddde86507ee0c115f05b281894a43d5f5abe385001b1735R7-R11))
*  Add `types.ts` file to define types for code generation process ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-19f27a7ccaf6ff3a0d9ead33a48b33e44301b3bcf5f0e9f9ce543f03b3855f48R1-R17))
*  Modify `CodeGeneratorSplitterService` class to use types from `types.ts` and set job status in Redis using `RedisService` ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-74301f42b1697fc3aaf4fcdcab83b98352343f406ba3482f31474ec81b7ff442L4-R16), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-74301f42b1697fc3aaf4fcdcab83b98352343f406ba3482f31474ec81b7ff442L29-R32), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-74301f42b1697fc3aaf4fcdcab83b98352343f406ba3482f31474ec81b7ff442L37-R41), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-74301f42b1697fc3aaf4fcdcab83b98352343f406ba3482f31474ec81b7ff442R47-R134), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-74301f42b1697fc3aaf4fcdcab83b98352343f406ba3482f31474ec81b7ff442L51-R155))
*  Modify `BuildRunnerService` class to pass `buildId` to `splitJobs` method and use `buildIdWithDomainName` in `copyFromJobToArtifact` method ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-a2d46b2df4132a6474b67758980206d93609e2deb576bab82bd20c13553f470bL26-R29), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-a2d46b2df4132a6474b67758980206d93609e2deb576bab82bd20c13553f470bL86-R118))
*  Modify `BuildRunnerController` class to import types from `types.ts` and change logic of `handleCodeGenerationSuccess` method ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-5c940f381b508c980f52d86bfc5418d7aa34d5c60f43628bfbd5a3459cf40d17R18), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-5c940f381b508c980f52d86bfc5418d7aa34d5c60f43628bfbd5a3459cf40d17L41-R91))
*  Add `redis.service.ts` file to define `RedisService` class that abstracts Redis operations ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-82eb4ed6b3077343a7651d4321c8c60f9c1140e6f43334d0984543439b949b1cR1-R26))
*  Modify `code-generator-splitter.service.spec.ts` file to update test cases for `CodeGeneratorSplitterService` class and mock `RedisService` dependency ([link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-e6b29c6a61d35d43aa885e01b9e2a4188cfad4dff781dcad2933f5b80091b851L10-R12), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-e6b29c6a61d35d43aa885e01b9e2a4188cfad4dff781dcad2933f5b80091b851R381-R382), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-e6b29c6a61d35d43aa885e01b9e2a4188cfad4dff781dcad2933f5b80091b851L387-R397), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-e6b29c6a61d35d43aa885e01b9e2a4188cfad4dff781dcad2933f5b80091b851L396-R410), [link](https://github.com/amplication/amplication/pull/7396/files?diff=unified&w=0#diff-e6b29c6a61d35d43aa885e01b9e2a4188cfad4dff781dcad2933f5b80091b851L410-R440))

## PR Checklist

- [ ] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-build-manager/src/build-logger/build-logger.controller.ts [hunk 1]
```diff
@@ -16,7 +17,7 @@ export class BuildLoggerController {
     @Body() logEntry: CodeGenerationLogRequestDto
   ): Promise<void> {
     const buildId = this.codeGeneratorSplitterService.extractBuildId(
-      logEntry.buildId
+      logEntry.buildId as JobBuildId<BuildId>
```
Comment (overbit) [inline packages/amplication-build-manager/src/build-logger/build-logger.controller.ts:20]: Try avoid casting in favour of a check within the function

File: packages/amplication-build-manager/src/build-runner/build-runner.service.spec.ts [hunk 2]
```diff
@@ -117,29 +170,526 @@ describe("BuildRunnerService", () => {
...
+
+      expect(spyOnCodeGeneratorSplitterServiceExtractBuildId).toBeCalledTimes(
+        1
+      );
+      expect(spyOnFsExtraCopy).toBeCalledWith(jobPath, artifactPath);
+      expect(result).toBe(true);
+      await expect(fsExtra.copy(jobPath, artifactPath)).resolves.not.toThrow();
+    });
+
+    it("should return false when it gets to the catch block", async () => {
```
Comment (overbit) [inline packages/amplication-build-manager/src/build-runner/build-runner.service.spec.ts:203]: Try to ask, what does the following test mean without looking at the code?

> copyFromJobToArtifact should return false when it gets to the catch block

File: packages/amplication-build-manager/src/build-runner/build-runner.service.spec.ts [hunk 3]
```diff
@@ -117,29 +170,526 @@ describe("BuildRunnerService", () => {
...
+      expect(spyOnCodeGeneratorSplitterServiceExtractBuildId).toBeCalledTimes(
+        1
+      );
+      expect(spyOnFsExtraCopy).toBeCalledTimes(1);
+      expect(result).toBe(false);
+    });
+  });
+
+  describe("runBuild", () => {
+    it("On code generation request, it should split the build into jobs when the comparison between the current version and the provided version is greater than or equal to 0", async () => {
```
Comment (overbit) [inline packages/amplication-build-manager/src/build-runner/build-runner.service.spec.ts:226]: Again here, the test title should be easy to understand without opening the code target of the test.
i.e.
_runBuild should split the build into jobs when code generator version has support for the new split job functionality_

Comment (overbit) [inline packages/amplication-build-manager/src/build-runner/build-runner.service.spec.ts:226]: Again here, the test title should be easy to understand without opening the code target of the test.
i.e.
```suggestion
    it("runBuild should split the build into jobs when code generator version has support for the new split job functionality", async () => {
```

File: packages/amplication-build-manager/src/build-runner/build-runner.service.spec.ts [hunk 4]
```diff
@@ -117,29 +170,526 @@ describe("BuildRunnerService", () => {
...
+        codeGeneratorVersion: expectedCodeGeneratorVersion,
+      });
+      expect(spyOnAxiosPost).toHaveBeenNthCalledWith(2, "http://runner.url/", {
+        resourceId: resourceId,
+        buildId: `${buildId}-${EnumDomainName.AdminUI}`,
+        codeGeneratorVersion: expectedCodeGeneratorVersion,
+      });
+    });
+
+    it("On code generation request, it should NOT split the build into jobs when the comparison between the current version and the provided version is less than 1", async () => {
```
Comment (overbit) [inline packages/amplication-build-manager/src/build-runner/build-runner.service.spec.ts:323]: Like above (but the opposite :) )

File: packages/amplication-build-manager/src/build-runner/build-runner.service.ts [hunk 5]
```diff
@@ -8,42 +8,255 @@ import { join, dirname } from "path";
...
-        throw new Error(error.message, {
-          cause: {
-            code: error.response?.status,
-            message: error.response?.data?.message,
-            data: error.config?.data,
-          },
+
+      // check if we need to split the build based on the code generator version.
+      // If the code generator version is greater than, or equal to the minDsgVersionToSplitBuild, we need to split the build
+      // otherwise, we can run the build without splitting it
```
Comment (overbit) [inline packages/amplication-build-manager/src/build-runner/build-runner.service.ts:56]: Let's avoid adding comment in favour of a clearer code.

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.spec.ts [hunk 6]
```diff
@@ -379,45 +379,171 @@ const onlyAdminInputJson: DSGResourceData = {
...
   });
 
-  it("should create two requests to start two jobs, one for admin by passing onlyAdminInputJson and one for server by passing onlyServerInputJson", () => {
-    const jobs = service.splitJobs(adminAndServerInputJson);
-    expect(jobs.length).toBe(2);
-    expect(jobs).toEqual([
-      [EnumDomainType.Server, onlyServerInputJson],
-      [EnumDomainType.AdminUI, onlyAdminInputJson],
-    ]);
+  it("should extract build id from jobBuildId", () => {
```
Comment (overbit) [inline packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.spec.ts:409]: What happens if the jobBuildId was "abc-bbb" (aka not valid)?

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.spec.ts [hunk 7]
```diff
@@ -379,45 +379,171 @@ const onlyAdminInputJson: DSGResourceData = {
...
+      expect(status).toBe(EnumJobStatus.InProgress);
+    });
   });
 
-  it("should build only admin, it will create one request to start a jobs with onlyAdminInputJson", () => {
-    const jobs = service.splitJobs(onlyAdminInputJson);
-    expect(jobs.length).toBe(1);
-    expect(jobs).toEqual([[EnumDomainType.AdminUI, onlyAdminInputJson]]);
+  describe("getJobStatus and setJobStatus", () => {
+    it("should get the job status", async () => {
```
Comment (overbit) [inline packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.spec.ts:520]: This test name doesn't make much sense:
> getJobStatus and setJobStatus should get the job status

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.spec.ts [hunk 8]
```diff
@@ -379,45 +379,171 @@ const onlyAdminInputJson: DSGResourceData = {
...
+      const spyOnRedisGet = jest
+        .spyOn(redisService, "get")
+        .mockResolvedValue(value);
+
+      const result = await service.getJobStatus(jobBuildId);
+      expect(spyOnRedisGet).toHaveBeenCalledWith(buildId);
+      expect(result).toBe(status);
+    });
+
+    it("should set the job status", async () => {
```
Comment (overbit) [inline packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.spec.ts:535]: same as before

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts [hunk 9]
```diff
@@ -1,17 +1,34 @@
...
-  AdminUI = "admin-ui",
-}
-
-type ResourceTuple = [EnumDomainType, DSGResourceData];
-
+type ResourceTuple = [JobBuildId<BuildId>, DSGResourceData];
 @Injectable()
 export class CodeGeneratorSplitterService {
-  splitJobs(dsgResourceData: DSGResourceData): ResourceTuple[] {
+  // TODO: rename to CodeGeneratorJobHandlerService (?)
```
Comment (overbit) [inline packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts:17]: Let's avoid adding redundant comments

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts [hunk 10]
```diff
@@ -1,17 +1,34 @@
...
+    private readonly redisService: RedisService,
+    private readonly logger: AmplicationLogger
+  ) {}
+
+  /**
+   * Splits the build into x jobs
+   * Currently we have 2 jobs: server and admin-ui
+   * Or only one if the user chose to build only one of them
+   */
+  async splitBuildsIntoJobs(
```
Comment (overbit) [inline packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts:28]: Let's avoid adding redundant comments. The function name is clear enough

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts (hunk 9 above)
Comment (abrl91) [reply packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts:17]: This is for me. I will remove it after we merge it to the base branch and rename the class name

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts [hunk 11]
```diff
@@ -26,31 +43,93 @@ export class CodeGeneratorSplitterService {
...
+    status: EnumJobStatus
+  ): Promise<void> {
+    const key = this.extractBuildId(jobBuildId);
+    const currentVal = await this.redisService.get<RedisValue>(key);
+    const newVal = {
+      ...currentVal,
+      [jobBuildId]: status,
+    };
+    this.logger.debug("Updating job status...", { newVal });
+    await this.redisService.set<RedisValue>(key, newVal);
```
Comment (overbit) [inline packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts:114]: We could improve performance and avoid race conditions by leveraging Redis Hash map to update part of the value stored.

Comment (overbit) [reply packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts:114]: We can open a new issue to improve this.

File: packages/amplication-build-manager/.env [hunk 12]
```diff
@@ -6,6 +6,16 @@ KAFKA_CLIENT_ID='build-manager'
...
+# Cache
+REDIS_HOST="127.0.0.1"
+REDIS_PORT=6379
+REDIS_PASSWORD=
+REDIS_USERNAME=
+REDIS_TLS_CA=
+REDIS_TLS_CERT=
+REDIS_TLS_KEY=
+REDIS_TLS_ENABLED="false"
+
```
Comment (overbit) [inline packages/amplication-build-manager/.env:18]: Is there any linked PR for the platform changes?

File: packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts (hunk 11 above)
Comment (abrl91) [reply packages/amplication-build-manager/src/code-generator/code-generator-splitter.service.ts:105]: https://github.com/amplication/amplication/issues/7418

**Review by overbit — APPROVED** [review]

**Pull request merged**